pcsx2/gui/Dialogs/GSDumpDialog.cpp

```diff
@@ -0,0 +1,776 @@
+			{
+				std::unique_ptr<char[]> fifo(new char[4]);
+				m_dump_file->Read(fifo.get(), 4);
+				GSData data = {id, std::move(fifo), 4, Dummy};
+				m_root_window->m_dump_packets.emplace_back(data);
+				break;
+			}
+			case Registers:
+			{
+				std::unique_ptr<char[]> regs_tmp(new char[8192]);
+				m_dump_file->Read(regs_tmp.get(), 8192);
+				GSData data = {id, std::move(regs_tmp), 8192, Dummy};
+				m_root_window->m_dump_packets.emplace_back(data);
+				break;
+			}
+		}
+	}
+
+	if (m_root_window->m_debug_mode->GetValue())
+		m_root_window->GenPacketList(m_root_window->m_dump_packets);
+
+	//return here to debug pre gs
+	//return;
+
+	GetCorePlugins().Init();
+	GSsetBaseMem((void*)regs);
+	if (GSopen2((void*)pDsp, renderer_override) != 0)
+	{
+		OnStop();
+		return;
+	}
+
+	GSsetGameCRC((int)crc, 0);
+
+	if (!GetCorePlugins().DoFreeze(PluginId_GS, 0, &fd, true))
+		m_running = false;
+	GSvsync(1);
+	GSreset();
+	GSsetBaseMem((void*)regs);
+	GetCorePlugins().DoFreeze(PluginId_GS, 0, &fd, true);
+
+	size_t i = 0;
+	size_t RunTo = 0;
+	size_t debug_idx = 0;
+
+	while (m_running)
+	{
+		if (m_root_window->m_debug_mode->GetValue())
+		{
+			if (m_root_window->m_button_events.size() > 0)
+			{
+				switch (m_root_window->m_button_events[0].index)
+				{
+					case Step:
+						if (debug_idx >= m_root_window->m_dump_packets.size())
+							debug_idx = 0;
+						RunTo = debug_idx;
+						break;
+					case RunCursor:
+						RunTo = m_root_window->m_button_events[0].index;
+						if (debug_idx > RunTo)
+							debug_idx = 0;
+						break;
+					case RunVSync:
+						if (debug_idx >= m_root_window->m_dump_packets.size())
+							debug_idx = 1;
+						auto it = std::find_if(m_root_window->m_dump_packets.begin() + debug_idx, m_root_window->m_dump_packets.end(), [](const GSData& gs) { return gs.id == Registers; });
+						if (it != std::end(m_root_window->m_dump_packets))
+							RunTo = std::distance(m_root_window->m_dump_packets.begin(), it);
+						break;
+				}
+				m_root_window->m_button_events.erase(m_root_window->m_button_events.begin());
+
+				if (debug_idx <= RunTo)
+				{
+					while (debug_idx <= RunTo)
+					{
+						m_root_window->ProcessDumpEvent(m_root_window->m_dump_packets[debug_idx++], regs);
+					}
+					auto it = std::find_if(m_root_window->m_dump_packets.begin() + debug_idx, m_root_window->m_dump_packets.end(), [](const GSData& gs) { return gs.id == Registers; });
+					if (it != std::end(m_root_window->m_dump_packets))
+						m_root_window->ProcessDumpEvent(*it, regs);
+
+					debug_idx--;
+				}
+
+				// do vsync
+				GSData vsync_end = {VSync, 0, 0, Dummy};
+				m_root_window->ProcessDumpEvent(vsync_end, regs);
+			}
+		}
+		else
+		{
+			while (i < (m_root_window->m_dump_packets.size()-1))
+			{
+				m_root_window->ProcessDumpEvent(m_root_window->m_dump_packets[i++], regs);
+
+				if (m_root_window->m_dump_packets[i].id == VSync)
+					break;
+			}
+			if (i >= m_root_window->m_dump_packets.size()-1)
+				i = 0;
```

Comment:
Unless this is intended, you're now essentially ignoring the last packet entirely by comparing to `size - 1` instead of just size.

I'm going to guess that the intention was to simplify this part from the original code:
> ```cs
>    if (gs_idx < dump.Data.Count && dump.Data[gs_idx].id == GSType.VSync)
>        break;
> ```

However, it looks to have been structured this way on purpose. As long as packets exist, we always want to process at least one. Only after the first packet do we check for VSync'd, but we can only do that after confirming that the next index is valid to begin with - thus requiring the `gs_idx < dump.Data.Count` check. If none of them have vsync, then every single one should get processed.

Now granted, it should've been
> ```cs
>    if (gs_idx >= dump.Data.Count || dump.Data[gs_idx].id == GSType.VSync)
>        break;
> ```

just so it doesn't end up checking the same false condition twice for an iteration that will fail.